addons/hr_recruitment/models/hr_applicant.py

```diff
@@ -397,6 +397,21 @@ def write(self, vals):
             for applicant in self:
                 if applicant.job_id.date_to:
                     applicant.candidate_id.availability = applicant.job_id.date_to + relativedelta(days=1)
+
+        # Because the company_id field is not tightly coupled between candidates
+        # we can end up in a situation where an application is in company A and
+        # candidate in company B. This results in recruitment users not being able
+        # to change anything on the application and bunch of tracebacks.
+        # Ideally the fields should be tightly coupled or hr.candidate should
+        # allow multiple companies to exist on it.
+        if vals.get("company_id"):
+            new_company = vals.get('company_id')
```

Comment:
Thank you! Yes that looks a lot nicer. Though you mean to update the candidate_id rather than the applicant_id right?(as this is all happening within the write method of the applicant)

something like this
```diff
         if vals.get("company_id"):
-            new_company = vals.get('company_id')
-            self.env["hr.candidate"].search(
-                [
-                    ("applicant_ids", "in", self.ids),
-                ]
-            ).write({"company_id": new_company})
+            self.candidate_id.write({"company_id": vals["company_id"]})
```